Refactor Virtualizer to improve performance, stability, and complexity

devongovett opened the pull request:
Fixes #6181, fixes #6074, fixes #5357, closes #5913, closes #1747

This is a significant refactor to `Virtualizer` to reduce its complexity, while improving performance and stability.

## Complexity

The current Virtualizer code originated over 9 years ago, before we even used React. While some of it has been updated, much of the original code survived. It supports a lot of features that we don't currently use, such as scroll anchoring, and has significant complexity for things like transactional animations. Things like scroll into view and focus management were also duplicated both in virtualizer and in other React Aria hooks, creating a tangled web of hard to maintain code.

This PR removes a lot of that complexity, reducing the main Virtualizer.ts from over 1200 lines down to 350. This is accomplished by removing support for animated updates, scroll anchoring, and other features we didn't use. It also removes scroll into view and focus management from Virtualizer, which can now simply use the same code we use for non-virtualized collections. This is possible because of the support for `persistedKeys` we added a while back, which ensures that the focused item is always persisted in the DOM even if it is scrolled out of view. This means that we can use the regular DOM scroll into view and focus management code instead of needing custom virtualizer-specific logic.

Animations and scroll anchoring may return in the future, but probably in a different form that is less virtualizer specific.

## Performance

This refactor also improves the performance of Virtualizer during scrolling. When it was initially designed, Virtualizer only handled single-level collections (e.g. lists). Over time it gained support for hierarchy, e.g. sections, or cells within rows. However, the data structures in the core virtualizer still operated on flattened collections. When reusing DOM nodes, this meant that a node representing a cell that was removed from view might be reused to render a cell in a different row. This "re-parenting" caused React to unmount and remount elements much more frequently than necessary, resulting in performance issues. We now maintain separate reuse queues per parent element, so that when a node is reused it is always placed within the same parent. This means that while scrolling, the React children do not change very much – they maintain the same keys and order, and only their contents change – which results in fewer DOM updates, browser layout and style invalidations, etc.

## Stability

Since the core of Virtualizer was originally written prior to React, it didn't follow React's rules in some places. For example, updating the collection or layout mutated the virtualizer object during render. This resulted in a side effect being called, triggering an additional render to update the visible views. In addition there were race conditions caused by some updates being triggered inside `requestAnimationFrame` to achieve batching. These caused some issues with React's StrictMode.

This PR refactors Virtualizer to run layout and updates during React's render lifecycle, rather than triggering additional re-renders as side effects. Updates to the visible rect, collection, layout, persisted keys, etc. all occur as regular React state updates and are passed into Virtualizer which performs some memoized computations and returns an array of visible views.

Virtualizer also now accepts a `layoutOptions` prop, which is passed into Layout objects to trigger updates. For example, we previously mutated layouts to update state like `isLoading`, which broke react's rules. Now, this can be passed via `layoutOptions`, and virtualizer handles triggering relayout during render when necessary.

devongovett commented on packages/@react-aria/focus/src/FocusScope.tsx:
```diff
@@ -715,6 +730,15 @@ function useRestoreFocus(scopeRef: RefObject<Element[]>, restoreFocus?: boolean,
+  // Dispatch a custom event that parent elements can intercept to customize focus restoration.
+  // For example, virtualized collection components reuse DOM elements, so the original element
+  // might still exist in the DOM but representing a different item.
+  if (node.dispatchEvent(new CustomEvent(RESTORE_FOCUS_EVENT, {bubbles: true, cancelable: true}))) {
```
This custom event fixes restoring focus when closing an ActionBar, which previously only worked because of the animation removing rows in TableView. Without the animation, row elements are reused to represent a different object immediately, and focus appears to be restored to the wrong item. Dispatching a custom event allows `useSelectableCollection` to pick this up and apply its own logic instead of the default provided by FocusScope.

devongovett commented on packages/@react-aria/selection/src/useSelectableCollection.ts:
```diff
@@ -293,6 +293,7 @@ export function useSelectableCollection(options: AriaSelectableCollectionOptions
   };
 
   // Store the scroll position so we can restore it later.
+  /// TODO: should this happen all the time??
```
I can't remember why we did this or what it fixed. Maybe it isn't needed anymore??

LFDanLu replied:
So you can reproduce by getting rid of the `scrollPos` references completely and by using the "useTable -> Scroll testing" story. Keyboard nav into the table and focus any cell that isn't in the first row, then tab out and shift tab back into the table:
Screen Recording 2024-05-29 at 5.05.56 PM

If my memory + notes on the PR serves, this is due to how focus lands on the last checkbox in the table before being marshalled to the proper tracked focused key -> causes a scroll position change in the scrollable body -> messes up scrollIntoView calculations. Think the same issue happens if shift tabbing to a previously focused table column as well

devongovett commented on packages/@react-spectrum/card/src/GridLayout.tsx:
```diff
@@ -130,41 +130,6 @@ export class GridLayout<T> extends BaseLayout<T> {
     );
   }
 
-  getVisibleLayoutInfos(rect) {
```
Handled by base class. Previously this implementation did not account for persisted keys.

devongovett commented on packages/@react-spectrum/list/src/ListView.tsx:
```diff
@@ -242,51 +240,17 @@ function ListView<T extends object>(props: SpectrumListViewProps<T>, ref: DOMRef
               if (type === 'item') {
-                return (
-                  <>
-                    {isListDroppable && collection.getKeyBefore(item.key) == null &&
```
Splitting lots of this out into separate components. The callback passed to virtualizer is memoized, so we can't rely on the values accessed from closure scope being correct here. Moving it into separate components and accessing what we need via context fixes this.

snowystinger reviewed (approved):
Did a bunch of focus testing and large collections + item addition/removal. Other than the ActionBar issue, I didn't find anything new.

snowystinger commented on packages/@react-spectrum/table/src/TableViewBase.tsx:
```diff
@@ -489,7 +448,27 @@ function TableViewBase<T extends object>(props: TableBaseProps<T>, ref: DOMRef<H
+        onFocusedResizer,
+        headerMenuOpen,
+        setHeaderMenuOpen,
+        renderEmptyState: props.renderEmptyState
```
o yeah, i know you aren't responsible for this ridiculous context, haha. I'm fairly certain that would be me ;)

majornista commented:
Can we fix https://github.com/adobe/react-spectrum/issues/4382 as well?

At present, we do not support sections in TableView; The rendered content within the `ScrollView` is `{state.visibleViews[1]}`, just the table body content.
I think we can fix [#4382](https://github.com/adobe/react-spectrum/issues/4382), by using `role="presentation"`  the `TableRowGroup` returned by the `renderWrapper` method at https://github.com/adobe/react-spectrum/blob/f1d3a8d353c4b5a41a463b08b19d69e5f7e1a1d8/packages/%40react-spectrum/table/src/TableViewBase.tsx#L294, and then replacing `role="presentation"` on the `ScrollView` at https://github.com/adobe/react-spectrum/blob/f1d3a8d353c4b5a41a463b08b19d69e5f7e1a1d8/packages/%40react-spectrum/table/src/TableViewBase.tsx#L620 with `role="rowgroup"` at (https://github.com/adobe/react-spectrum/blob/f1d3a8d353c4b5a41a463b08b19d69e5f7e1a1d8/packages/%40react-spectrum/table/src/TableViewBase.tsx#L638. We can add a comment explaining this:
```tsx
            /*
              Firefox and Chrome make generic elements using CSS overflow 'scroll' or 'auto' tabbable,
              including them within the accessibility tree, which breaks the table structure in Firefox,
              Using tabIndex={-1} prevents the ScrollView from being tabbable, and using role="rowgroup"
              and role="presentation" on the table body content fixes the table structure.
            */
            role="rowgroup"
            tabIndex={isVirtualDragging ? null : -1}
  ```

devongovett commented:
Added some additional performance improvements found while profiling an example project from Workfront.

1. Attempt to keep the number of rows in the DOM consistent at all times. Removing and re-adding rows has a huge performance cost. Previously this would change due to rounding - when rows were perfectly aligned with the viewport, there would be fewer than when they were not aligned. Now we always round the y position down and the height up to the nearest row boundaries.
2. Adding on to #1, our overscanning logic attempted to use a rolling average of the amount of time needed for one render cycle to estimate how much extra to render. However, this resulted in the overscan amount changing during scrolling, causing rows to be removed and added more than necessary. We now use a fixed amount of 1/3 in the direction the user is scrolling to keep the number consistent.
3. Reuse queues should be FIFO – when reusing views we want to preserve the order they were in before. Reordering DOM nodes is expensive.
4. Unrelated, but made a difference: reading `navigator.userAgentData` is expensive, so our platform checks are now cached.

This together with the previous optimizations resulted in a huge improvement on the added storybook example: average per-frame (scroll event) render time went from ~70ms to ~4ms. The example has 10,000 rows and 50 columns, with complex cell content such as tags, status lights, links, buttons, etc., and it now scrolls smoothly with no dropped frames.

snowystinger reviewed (approved):
Even in dev strict that table works pretty well
one odd thing I noticed, the progress bars seem to be reused, so they animate whenever they scroll into view

snowystinger commented:
Found a little more odd behavior, as I keyboard navigate rows, the tag group seems to also navigate?
https://github.com/adobe/react-spectrum/assets/698229/5ca203d9-9757-430d-b2f9-99554c40183c

devongovett commented:
> Found a little more odd behavior, as I keyboard navigate rows, the tag group seems to also navigate?

I don't think that's new right?

LFDanLu commented on packages/@react-stately/virtualizer/src/OverscanManager.ts:
```diff
@@ -50,42 +34,21 @@ export class OverscanManager {
-      overscanned.y -= overscanY * 0.2;
-      overscanned.height += overscanY + overscanY * 0.2;
-    } else {
+    let overscanY = this.visibleRect.height / 3;
+    overscanned.height += overscanY;
```
~~Curious about the math here, is this essentially signalling to the virtualizer that the available rect is an additional 1/3 of the visible rect bounds in the direction of the scrolling~~ Just saw you additional comment. Was the 1/3 just a number that felt reasonable performance wise? Is there any downside to returning 0 for overscanY if the velocity is 0 (aka the user isn't scrolling)? I suppose its nice to always have the extra rows rendered at all times and to avoid changing this overscan value even more

LFDanLu commented on packages/@react-spectrum/table/src/TableViewBase.tsx:
```diff
@@ -278,25 +277,21 @@ function TableViewBase<T extends object>(props: TableBaseProps<T>, ref: DOMRef<H
 
   // This overrides collection view's renderWrapper to support DOM hierarchy.
   type View = ReusableView<GridNode<T>, ReactNode>;
-  let renderWrapper = (parent: View, reusableView: View, children: View[], renderChildren: (views: View[]) => ReactElement[]) => {
-    let style = layoutInfoToStyle(reusableView.layoutInfo, direction, parent && parent.layoutInfo);
-    if (style.overflow === 'hidden') {
-      style.overflow = 'visible'; // needed to support position: sticky
-    }
-
```
Another thing I found was the focus rings in the hide header are cut off now, doesn't happen on main:
<img width="443" alt="image" src="https://github.com/adobe/react-spectrum/assets/9661127/a9145644-8d83-4f88-b8ac-362aad79ced7">

Same root issue with the cellWrapper not having `overflow: 'visible'`

LFDanLu reviewed (approved):
Approving for testing, things seemed to work well in general and the performance is definitely a lot better. Just one small bug I found

LFDanLu commented:
Oh one more bug I found, but it seems that focus is being lost to the body now when you delete the last row in the collection in the CRUD stories: https://reactspectrum.blob.core.windows.net/reactspectrum/9affd262041afc552ae817be3568021d1af9168c/storybook/index.html?path=/story/tableview--crud&providerSwitcher-express=false
https://reactspectrum.blob.core.windows.net/reactspectrum/9affd262041afc552ae817be3568021d1af9168c/storybook/index.html?path=/story/listview--remove-list-items&providerSwitcher-express=false

I suppose this might be something to do with the FocusScope changes and/or the the removal of the animations perhaps? It only happens for delete operations that happen in a overlay and not for inline deletion like in https://reactspectrum.blob.core.windows.net/reactspectrum/9affd262041afc552ae817be3568021d1af9168c/storybook/index.html?path=/story/tableview--inline-delete-buttons&providerSwitcher-express=false.

https://github.com/adobe/react-spectrum/blob/9affd262041afc552ae817be3568021d1af9168c/packages/%40react-aria/selection/src/useSelectableCollection.ts#L427-L429 doesn't seem to trigger in the buggy cases since the focusedKey is already set to `null`?

devongovett commented:
Merging for testing.

piecyk commented on packages/@react-aria/selection/src/useSelectableCollection.ts:
```diff
@@ -342,7 +343,7 @@ export function useSelectableCollection(options: AriaSelectableCollectionOptions
     }
 
-    if (!isVirtualized && manager.focusedKey != null) {
+    if (manager.focusedKey != null) {
```
Was removing !isVirtualized intentional here?

devongovett replied:
yeah we centralized the logic between both virtualized and non-virtualized collections so now they are both handled here. was there a concern or just curious?
